File: problem-solving/baekjoon/2960_에라토스테네스의_체.go
// https://www.acmicpc.net/problem/2960
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n, k int
	fmt.Fscanln(reader, &n, &k)
	fmt.Fprintln(writer, eratos(n, k))
}

func eratos(n, k int) (kth int) {
	primeNumbers := initPrimeNumbers(n)
	for i := 2; i < n+1; i++ {
		if primeNumbers[i] {
			for j := i; j < n+1; j += i {
				if primeNumbers[j] {
					primeNumbers[j] = false
					k--
					if k == 0 {
						kth = j
						return
					}
				}
			}
		}
	}
	return
}

func initPrimeNumbers(n int) []bool {
	var primeNumbers = make([]bool, n+1)
	for i := range primeNumbers {
		primeNumbers[i] = true
	}
	return primeNumbers
}

File: problem-solving/baekjoon/2960_에라토스테네스의_체.py
# https://www.acmicpc.net/problem/2960
import sys

def eratos(n, k):
    prime_numbers = [True for _ in range(n+1)]
    kth = 0
    for i in range(2, n+1):
        if prime_numbers[i]:
            for j in range(i, n+1, i):
                if prime_numbers[j]:
                    prime_numbers[j] = False
                    k -= 1
                    if k == 0:
                        kth = j
                        return kth
    return kth

if __name__ == "__main__":
    n, k = list(map(int, sys.stdin.readline().split()))
    print(eratos(n, k))